### DISCUSSION TITLE
How to SetDirtyState in a SourceModelStorage

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1000] [Q&A] [answered]
Hi,

I have custom `SourceModelStorage` implementation where  I need to read content from different file locations.
It may happen that in the call of `loadSourceModel` I recognize that some parts of my model need to be updated because the content of some linked files has changed. So this affects the overall state in my model during loading.

Now my question is: how can I mark the model immediately after loading as 'dirty' ?

I already have seen that there is an action class `SetDirtyStateAction`.

**Comment by rsoika** [comment, marked as answer]
I found the solution by myself and so I just post my code example which may help other users:


```java
public class MySourceModelStorage implements SourceModelStorage {

    @Inject
    protected ActionDispatcher actionDispatcher;
    @Override
    public void loadSourceModel(final RequestModelAction action) {

        // Load your model....
        ....
        .......
       if (externalFileContenthasChanged) {
          // now ....SET IS DIRTY flag...
          SetDirtyStateAction dirtyAction = new SetDirtyStateAction();
          dirtyAction.setDirty(true);
          dirtyAction.setReason("Updated linked File Content");
          // dispatch new DirtyState....
          actionDispatcher.dispatchAfterNextUpdate(dirtyAction);
       }
      ....
    
    }
```

After I created my `GModelState` and some content has changed, I immediately set the model as Dirty by dispatching a new `SetDirtyStateAction`.

**Reply by tortmayr** [reply]
Thanks for sharing your solution. This definitely will be helpful for other users as well.
Just as an FYI: You can also pass the dirty & reason values directly in the constructor of the `SetDirtyStateAction`.

**Reply by rsoika** [reply]
@tortmayr sorry but I have to come back to this with a new question:
What could be the reason that Actions like the `SetDirtyStateAction` or `ServerMessageAction` are not working in VS-Code ? 

Did I miss to add some additional implementation? Or do you expect that the code should work in VS-Code too?

I debugged this a little bit. In the Browser Console, I can see that both a isDirty as also the serverMessage messages are received. But Neither one nor the other event has the expected effect 

![image](https://github.com/eclipse-glsp/glsp/assets/1339124/89a79aba-bed4-4d75-b5e5-1c0e3455e610)

**Reply by tortmayr** [reply]
Hi @rsoika,

> @tortmayr sorry but I have to come back to this with a new question: What could be the reason that Actions like the `SetDirtyStateAction` or `ServerMessageAction` are not working in VS-Code ?
> 
> Did I miss to add some additional implementation? Or do you expect that the code should work in VS-Code too?

There seems to be a bug in the VSCode-integration that causes incorrect handling of some `SetDirtyStateAction`. I have opened https://github.com/eclipse-glsp/glsp/issues/1006 to track this.
Regarding the serverMessage actions: It looks like this feature is currently not implemented (see also https://github.com/eclipse-glsp/glsp/issues/205).


> I found the solution by myself and so I just post my code example which may help other users:
> 
> ```java
> public class MySourceModelStorage implements SourceModelStorage {
> 
>     @Inject
>     protected ActionDispatcher actionDispatcher;
>     @Override
>     public void loadSourceModel(final RequestModelAction action) {
> 
>         // Load your model....
>         ....
>         .......
>        if (externalFileContenthasChanged) {
>           // now ....SET IS DIRTY flag...
>           SetDirtyStateAction dirtyAction = new SetDirtyStateAction();
>           dirtyAction.setDirty(true);
>           dirtyAction.setReason("Updated linked File Content");
>           // dispatch new DirtyState....
>           actionDispatcher.dispatchAfterNextUpdate(dirtyAction);
>        }
>       ....
>     
>     }
> ```
> 
> After I created my `GModelState` and some content has changed, I immediately set the model as Dirty by dispatching a new `SetDirtyStateAction`.

While this approach seems to work at first glance, I see one problem with it. By directly dispatching the `SetDirtyStateAction` you bypass the `CommandStack`. In the default implementation the ModelState [reuses the internal dirty state model of the `CommandStack`](https://github.com/eclipse-glsp/glsp-server/blob/fb07a83b25e227c078c10b96b586ff7e19cb1220/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/model/DefaultGModelState.java#L120). This means with your approach the intial dirty information can get lost after an undo e.g:

1. Model loading & external file change -> dirty: true
2. Move a node -> dirty:true
3. Undo move -> dirty:false (the commandstack is empty => Server sends a non-dirty event to the client, but it should still be dirty)

To prevent this I would recommend to use a different approach and execute a custom `SetDirtyCommand` on the commandStack.
The custom command can be very simple and is more or less a `NoOp` Command that cannot be undone:
```java
import org.eclipse.emf.common.command.AbstractCommand;

public class SetDirtyCommand extends AbstractCommand {

   @Override
   public void execute() {}

   @Override
   protected boolean prepare() {
      return true;
   }

   @Override
   public boolean canUndo() {
      return false;
   }

   @Override
   public void redo() {}
}
```

Then you could update your `SourceModelStorage` as follows:
```java
public class MySourceModelStorage implements SourceModelStorage {

    @Inject
    protected ActionDispatcher actionDispatcher;
    @Override
    public void loadSourceModel(final RequestModelAction action) {

        // Load your model....
        ....
        .......
       if (externalFileContenthasChanged) {
           modelState.execute(new SetDirtyCommand());
          // For Theia the line above is enough, if you want to bypass the current VS Code bug, you have to dispatch an additional `SetDirtyStateAction` with reason "operation" here. eg.:
                this.actionDispatcher.dispatch(new SetDirtyStateAction(modelState.isDirty(), SetDirtyStateAction.Reason.OPERATION));
       }
      ....
    
    }
```

**Reply by rsoika** [reply]
Thanks @tortmayr your workaround with the DirtyFlag. It works perfect - also now in VSCode. Thanks again.

With the missing notification alert in VS-Code I will wait for [Pull Reqeuest #35](https://github.com/eclipse-glsp/glsp-vscode-integration/pull/35)
